### ISSUE TITLE
Experiencing unintentional "jumping" of text

### ISSUE THREAD

**Issue by DavidBjelland** [Zettlr/Zettlr#1621]
## Description

I'm frequently experiencing unintentional "jumping" of text when working on it.
This behavior occurs especially frequent when formatting/rendering is used (e.g. markdown formatting or external links). 

## Reproducing

The behavior occurs whenever working on text. See also the attached video (no scrolling on my end involved).

## Platform
 - OS and version: Windows 10 Pro 1909
 - Zettlr Version: 1.8.5

## Additional information

https://user-images.githubusercontent.com/63207312/104181082-7a1e1800-540e-11eb-9481-f9925196cc4f.mp4

**Comment by crbald** [comment]
I also frequently have these jumps but haven't figured out the exact reason either which makes reliable reproduction a bit hard for me.

**Comment by nathanlesage** [comment]
I've noticed these as well, but I haven't found the cause within the code base yet …

**Comment by DavidBjelland** [comment]
The only thing I can add is that I've had this behavior in previous versions of Zettlr.
As far as I remember it occured in version 1.7.2 as well.

**Comment by ecodiv** [comment]
> 
> 
> The only thing I can add is that I've had this behavior in previous versions of Zettlr.
> As far as I remember it occured in version 1.7.2 as well.

I have not experienced this in prior versions, it started with 1.8.5 as far as I can tell. First thought this was because by change just after updating, I started working with a file with many images, and sometimes it takes a bit of time to 'load' all images. But after reverting back to 1.8.4 it doesn't happen.

**Comment by DavidBjelland** [comment]
The issue remains the same in version 1.8.6, the "jumping" seems to be faster now.

**Comment by nathanlesage** [comment]
Found the bug, fix is in progress

**Comment by DavidBjelland** [comment]
Thank you for your work! I'm excited for the next version.

**Comment by BellLongworth** [comment]
Just installed 1.8.7 and things are still jumping around occasionally. The cursor is not at the end of the page.
(Windows 10)

**Comment by nathanlesage** [comment]
@BellLongworth This pertains most likely to the Markdown commands. I haven't had time to dive into that as well. I just fixed unintentional jumping with regard to resizing headings via styles. The Markdown commands are also on my list, and I will get to that, but it may take some more time.

**Comment by jasonhbartlett** [comment]
Hi, just jumping in here to say I also have this exact same issue.  (Ver. 1.8.7) I have a file that I jot down my daily work in a task list like this:

##### Jan 31st - Feb 5th

Sunday, Jan 31st
    - [ ] Worked on XYZ
    - [ ] Worked on ABC
    - [ ] etc

Monday, Feb 1st
    - [ ] Worked on MNO
    - [ ] Worked again on ABC

I've been doing it for about 6 months. The file is about 600 lines (4100 words) so far.

I notice the jumps happen when I stop typing and pause for a second or two.  It's taking what I'm currently writing and moving it below where I can't see it unless I scroll up again.

**Comment by BellLongworth** [comment]
On some of my documents this happens so often to be almost infuriating... :(

**Comment by BellLongworth** [comment]
I managed to isolate this - happens whenever the file autosaves a few seconds after a change.  Happens even in a simple file with no markup, just copy-pasting the same line for enough times to make the scroll bar appear.

**Comment by jasonhbartlett** [comment]
Well, yeah this is super annoying.  Wish I could turn off auto-save then.  :(

**Comment by jweickm** [comment]
Just wanted to say that I have the same issue. But I am glad that it has already been reported and is being looked at.

**Comment by nathanlesage** [comment]
> Wish I could turn off auto-save then. :(

Don't you worry, it'll be off by default in 2.0

Plus, I am confident that the issue will be gone in the new 2.0 line. But let's wait for the betas!

**Comment by BellLongworth** [comment]
No chance for an earlier fix? Things are jumping every few seconds. I actually started to write in Notepad++ and then copy the text to Zettlr once in a while...

**Comment by canpolat** [comment]
> I managed to isolate this - happens whenever the file autosaves a few seconds after a change. Happens even in a simple file with no markdown, just copy-pasting the same line for enough times to make the scroll bar appear.

This is in alignment with my experience. I save my document manually and I see the jump after each save. In the below screenshot I first copy a few paragraphs of text (no markdown) and then save the document manually (Ctrl + S).

![2021-02-27_13-59-18](https://user-images.githubusercontent.com/459891/109387885-0ffcfc00-7904-11eb-8ceb-4d8e7ccd596e.gif)

**Comment by nathanlesage** [comment]
> No chance for an earlier fix?

I don't think so :( I'm completely booked out with rewriting 2.0, and I have not been able to precisely isolate this ni 1.8.7, so there's not guarantee I will be able to find the problem …

**Comment by nathanlesage** [comment]
I think I just found the culprit – there were some `refresh`-statements in the markdown commands which seem to have caused the jumping. Jumping after saving does not occur anymore in the current 2.0 branch. So, I guess another mystery solved!

**Comment by OldDon** [comment]
Hi there. I am brand new to Zettlr but I just wanted to pass on my appreciation for your, hmm, AWESOMENESSESSess! Do you think this could be why I can't earn a living as a writer? Seriously, I am amazed at how quickly you have resolved this especially as it seemed like a random problem.

I hope you are well and keeping safe.

Many thanks,

Newbie

**Comment by Serenade600** [comment]
> I think I just found the culprit – there were some `refresh`-statements in the markdown commands which seem to have caused the jumping. Jumping after saving does not occur anymore in the current 2.0 branch. So, I guess another mystery solved!

Since the reason causing auto jumping after saving has been detected, can we expect an early fixed version? From my perspective, the jumping heavily disrupts my thought when typing as I use ctrl+s to save very often. This one bug obscures all other brilliant features and nearly makes me quit this wonderful application.

**Comment by nathanlesage** [comment]
@Serenade600 That is completely understandable. However, due to time-constraints, I can't do backports. But I hope to begin beta with the 2.0 branch quickly … the 2.0-dev branch is very stable already.

**Comment by pstadler1990** [comment]
Just wanted to report that the jumping also occurs on the new 2.0 update when saving the document.

Version: 2.0.0-beta.1

**Comment by BellLongworth** [comment]
Bummer. I was so expecting to finally work. At this point I'm writing long stuff in notepad and copying to Zettlr. Would've gave up on any other tool long ago, still believe in Zettlr.

**Comment by nathanlesage** [comment]
@pstadler1990 @BellLongworth Would ya mind recording your screen to show how that happens? Because I have not experienced this in the past half year o.O

**Comment by pstadler1990** [comment]
https://user-images.githubusercontent.com/20464813/129325268-5e24dad3-3914-4cc2-a3d8-d8846c1a317a.mp4

**Comment by nathanlesage** [comment]
That is super weird, on macOS this isn't happening o.O I have to test this tomorrow on my Windows machine.

**Comment by nathanlesage** [comment]
That is super weird, on macOS this isn't happening o.O I have to test this tomorrow on my Windows machine.

**Comment by pstadler1990** [comment]
I have another document where the effect is even stronger (as in "more jumping"), but I can't currently show it on public.

**Comment by BellLongworth** [comment]
That's exactly what I see as well. Windows 10.

**Comment by BellLongworth** [comment]
I just now installed beta-1 and the issue seems to be gone there for me. I did notice that there is no periodic autosave anymore so maybe it is related. When I save manually there is now no jumping. In the last stable version there was jumping at both manual save and autosave.

**Comment by nathanlesage** [comment]
> I did notice that there is no periodic autosave anymore so maybe it is related. When I save manually there is now no jumping.

That was what I originally hoped for with my code-changes!

**Comment by BellLongworth** [comment]
@nathanlesage with autosave restored in beta-2 I still don't have the jumping behavior, for both auto and manual saves. I'm happy now. :)

**Comment by nathanlesage** [comment]
That's nice to hear! What about @pstadler1990 ?

**Comment by pstadler1990** [comment]
I'm sorry, but the problem still occurs even with the most recent version. It still seems to appear mostly when saving manually.

**Comment by BellLongworth** [comment]
And... it's back!
Both for saving manually and autosave. Started happening in a certain file once it became large enough (~2.7k words). The jumping back is not as big (in # of lines) as previously though.

**Comment by BellLongworth** [comment]
I think I found a hint - this is really size dependent.
What I have:
13270 chars - all OK.
13271 chars - small jump when editing at the end of the document. Big one when editing at the middle of it.

Reproduced on one document. 
(beta-3, windows 10)

**Comment by BellLongworth** [comment]
Progress!

This is completely correlated to the Frequent "modified remotely" message bug!

In removing all markdown from the document the issue still reproduces, when character number 13480 is added. When I unchecked the "always load remotely" checkbox that character was exactly when BOTH issues started reproducing. Whenever I chose to load the file remotely, the jump happened. If I chose to not load the file, everything was ok.

Some more information: With the "always load remotely" checked I noticed that when I scroll up or down after saving some of the headings that get in the view get re-rendered. So a level 4 render first appears as smallish and then gets rendered to its full size. This leads to some more jumping. 

Bottom line - I think this is related to redrawing the file after the "remote load" thing.

**Comment by pbeltran** [comment]
I confirm the jumping with beta-3, windows.

**Comment by pstadler1990** [comment]
> 
> 
> Progress!
> 
> This is completely correlated to the Frequent "modified remotely" message bug!
> 
> After removing all markdown from the document the issue still reproduces, when character number 13480 is added. When I unchecked the "always load remotely" checkbox that character was exactly when BOTH issues started reproducing. Whenever I chose to load the file remotely, the jump happened. If I chose to not load the file, everything was ok.
> 
> One more hint: With the "always load remotely" checked I noticed that when I scroll up or down after saving some of the headings that get in the view get re-rendered. So a level 4 render first appears as smallish and then gets rendered to its full size. This leads to some more jumping.
> 
> Bottom line - I think this is related to redrawing the file after the "remote load" thing.

I'm using a mix of markdown and latex in my "jumping" document. But even though it's all just plain text, right?

**Comment by nathanlesage** [comment]
> I'm using a mix of markdown and latex in my "jumping" document. But even though it's all just plain text, right?

Yes it is.

You know what I just realized? There is a second instance when the document might look as if it's jumping: When Zettlr renders some elements that actually change the height of lines (such as big headings or tables or images), and these are only rendered when you scroll them into view. This could also be an issue, because I'm almost certain that most of you don't actually have any problems with file reloading (you can simply test this by having Zettlr prompt you whenever a file _actually_ changes remotely).

**Comment by BellLongworth** [comment]
Yep. I saw that happening (see a few comments above).

For the record, the new beta did not solve my jumping problem. Not sure if it was intended to.

**Comment by noah-arroyo** [comment]
Same. Jumping in previous and current beta.

**Comment by nathanlesage** [comment]
Could someone maybe do a screencast of what is happening? I still think that it's related to all elements re-rendering after Zettlr swaps the "old" content with the "new" one, but I'm running out of ideas here.

And I just noticed that I don't log anything, which I'll change now, so that we have an indicator in the logs that indeed the file is being exchanged.

**Comment by nathanlesage** [comment]
And just did so. If anyone wants to test it out: I'm going to trigger a nightly build right away. Then you could see if whenever your documents appear to jump, if a remote change message is logged for said file (level is "info", so the blue messages). If so, please report back!

**Comment by noah-arroyo** [comment]
For me, the behavior is unpredictable. But I'll do my best to record it.

I've just downloaded and installed the latest nightly build. You're saying that this build logs remote changes? Or you're saying that you'll put that functionality into the _next_ nightly build?

**Comment by noah-arroyo** [comment]
@nathanlesage check Zettlr's Twitter account for a dm from me, regarding this bug.

**Comment by nathanlesage** [comment]
@noah-arroyo Thank you! I have already replied to you. I now think that my initial suspicion was actually true that certain cloud providers in fact do tamper with the files, and in your case the app just blatantly overwrote the newer file with an older revision which is just outrageous :D 

As I wrote in my reply, I think one remidy might be to just ignore change events if the time stamp of that event lies sometime in the past. Right now Zettlr checks if the modification time of the change event is THE SAME as the internal descriptor, but we could change that.

I think I might add an experimental option for that, because we still don't know if OneDrive does exactly the same as the box app, or not. I'll keep you posted!

**Comment by nathanlesage** [comment]
Sooo I just added the experimental option to the advanced options. I will NOW trigger a new nightly. Check nightly.zettlr.com to see when it's done. As soon as it is, please try it out and everyone who experiences the jumping bug activate that "EXPERIMENTAL" option in the advanced preferences.

Please say if this makes the bug disappear!

**Comment by BellLongworth** [comment]
I don't see that option in Preferences/Advanced. What am I doing wrong? (installed nightly)
![image](https://user-images.githubusercontent.com/63842998/134546870-0a028c2e-0c75-4006-94d1-153c8d3a6016.png)

**Comment by noah-arroyo** [comment]
@nathanlesage I've just downloaded the nightly, checked that box in the advanced settings and will report back.

@BellLongworth must be working with a different OS? (I'm on Windows 10, and downloaded Zettlr-2.0.0-beta.4-x64.exe.)

**Comment by BellLongworth** [comment]
@noah-arroyo that's the same file I downloaded.
Looking at the timestamp in the Windows digital signatures window it looks like yesterday's build. Weird. I'll try downloading again, for some reason the download speed is very slow and it takes 30 minutes to finish.
![image](https://user-images.githubusercontent.com/63842998/134552029-34b6a7ad-197b-4378-b50e-8ea1e415856f.png)

**Comment by BellLongworth** [comment]
OK. I probably originally downloaded the file a bit too fast, before the new one finished uploading or something.

@nathanlesage  Good news - today's nightly solved the jumping problem for me, for both manual and auto save.

**Comment by noah-arroyo** [comment]
Text appears to be steady for me too. Been working in Zettlr for over an hour, no jumping.

:)

**Comment by nathanlesage** [comment]
> OK. I probably originally downloaded the file a bit too fast, before the new one finished uploading or something.

Yeah when I say "I trigger it NOW" it usually takes about an hour or so to actually upload the nightly to the server – so that's why I added the timestamps for the file so that it would be easier to discern "Is this still the nightly from two days ago or from today?". In any case, you can always check the publicly visible Build action on the "Actions" tab: Once that shows a green light it's ready!

***

But then: I'm really happy that it's working! That's what I had hoped for!

I also had that setting activated to test how it would behave in "normal" circumstances and I couldn't see any problems, so I'll just remove that option and hardcode it!

**To everyone else who hasn't yet responded**: I'll close this issue with said commit because I think two confirmations sound pretty good, BUT if that should not work in your case, please feel free to comment and indicate that my commit later today doesn't fix it for you, but I hope it does!
